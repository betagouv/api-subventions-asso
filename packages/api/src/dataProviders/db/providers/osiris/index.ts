import osirisRequestPort from "./osiris.request.port";
import osirisActionPort from "./osiris.action.port";

export { osirisRequestPort, osirisActionPort };
