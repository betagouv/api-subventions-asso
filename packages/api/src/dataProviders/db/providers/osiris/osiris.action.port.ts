import { FindOneAndUpdateOptions } from "mongodb";
import { MongoCnxError } from "../../../../shared/errors/MongoCnxError";
import OsirisActionEntity from "../../../../modules/providers/osiris/entities/OsirisActionEntity";
import OsirisActionEntityDbo from "../../../../modules/providers/osiris/entities/OsirisActionEntityDbo";
import MongoPort from "../../../../shared/MongoPort";
import Siren from "../../../../valueObjects/Siren";
import OsirisActionAdapter from "./osirisAction.adapter";

export class OsirisActionPort extends MongoPort<OsirisActionEntityDbo> {
    collectionName = "osiris-actions";

    async createIndexes() {
        await this.collection.createIndex({ "indexedInformations.uniqueId": 1 }, { unique: true });
        await this.collection.createIndex({ "indexedInformations.osirisActionId": 1 });
        await this.collection.createIndex({ "indexedInformations.requestUniqueId": 1 });
        await this.collection.createIndex({ "indexedInformations.compteAssoId": 1 });
        await this.collection.createIndex({ "indexedInformations.siret": 1 });
    }

    // Action Part
    public async add(osirisAction: OsirisActionEntity) {
        await this.collection.insertOne(OsirisActionAdapter.toDbo(osirisAction));
        return osirisAction;
    }

    /*
     * @deprecated
     * */
    public async update(osirisAction: OsirisActionEntity) {
        const options: FindOneAndUpdateOptions = { returnDocument: "after", includeResultMetadata: true };
        const { _id, ...actionWithoutId } = OsirisActionAdapter.toDbo(osirisAction);
        const updateRes = await this.collection.findOneAndUpdate(
            { "indexedInformations.uniqueId": osirisAction.indexedInformations.uniqueId },
            { $set: actionWithoutId },
            options,
        );

        //@ts-expect-error -- mongo typing expects no metadata
        const dbo = updateRes?.value;
        if (!dbo) throw new MongoCnxError();
        return OsirisActionAdapter.toEntity(dbo);
    }

    public upsertOne(osirisAction: OsirisActionEntity) {
        const options = { upsert: true } as FindOneAndUpdateOptions;
        const { _id, ...actionWithoutId } = osirisAction;
        return this.collection.updateOne(
            { "indexedInformations.uniqueId": osirisAction.indexedInformations.uniqueId },
            { $set: actionWithoutId },
            options,
        );
    }

    public async bulkUpsert(osirisActions: OsirisActionEntity[]) {
        const bulk = osirisActions.map(a => {
            const { _id, ...actionWithoutId } = a;
            return {
                updateOne: {
                    filter: { "indexedInformations.uniqueId": a.indexedInformations.uniqueId },
                    update: { $set: actionWithoutId },
                    upsert: true,
                },
            };
        });
        return bulk.length ? this.collection.bulkWrite(bulk, { ordered: false }) : Promise.resolve();
    }

    public async findByUniqueId(uniqueId: string) {
        const dbo = await this.collection.findOne({
            "indexedInformations.uniqueId": uniqueId,
        });
        if (!dbo) return null;
        return OsirisActionAdapter.toEntity(dbo);
    }

    public cursorFind(query = {}) {
        return this.collection.find(query);
    }

    public async findByRequestUniqueId(requestUniqueId: string) {
        const dbos = await this.collection.find({ "indexedInformations.requestUniqueId": requestUniqueId }).toArray();
        return dbos.map(dbo => OsirisActionAdapter.toEntity(dbo));
    }

    public async findBySiren(siren: Siren) {
        const dbos = await this.collection
            .find({ "indexedInformations.siret": new RegExp(`^${siren.value}\\d{5}`) })
            .toArray();
        return dbos.map(dbo => OsirisActionAdapter.toEntity(dbo));
    }
}

const osirisActionPort: OsirisActionPort = new OsirisActionPort();

export default osirisActionPort;
