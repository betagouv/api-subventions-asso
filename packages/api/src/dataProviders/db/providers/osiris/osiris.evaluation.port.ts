import { FindOneAndUpdateOptions } from "mongodb";
import OsirisEvaluationEntity from "../../../../modules/providers/osiris/entities/OsirisEvaluationEntity";
import MongoRepository from "../../../../shared/MongoRepository";
import Siret from "../../../../valueObjects/Siret";
import Siren from "../../../../valueObjects/Siren";

export class OsirisEvaluationPort extends MongoRepository<OsirisEvaluationEntity> {
    readonly collectionName = "osiris-evaluation";

    public async findByActionId(actionId: string) {
        return this.collection.findOne({
            "indexedInformations.osirisActionId": actionId,
        });
    }

    public async add(request: OsirisEvaluationEntity) {
        await this.collection.insertOne(request);
        return request;
    }

    public async update(request: OsirisEvaluationEntity) {
        const options = { returnNewDocument: true, includeResultMetadata: true } as FindOneAndUpdateOptions;
        const { _id, ...requestWithoutId } = request;
        return (await this.collection.findOneAndUpdate(
            { "indexedInformations.osirisActionId": request.indexedInformations.osirisActionId },
            { $set: requestWithoutId },
            options,
            //@ts-expect-error -- mongo typing expects no metadata
        ))!.value as OsirisEvaluationEntity;
    }

    public findsBySiret(siret: Siret) {
        return this.collection
            .find({
                "legalInformations.siret": siret.value,
            })
            .toArray();
    }

    public findBySiren(siren: Siren) {
        return this.collection
            .find({
                "legalInformations.siret": new RegExp(`^${siren.value}\\d{5}`),
            })
            .toArray();
    }

    public cursorFind(query = {}) {
        return this.collection.find(query);
    }

    async createIndexes() {
        await this.collection.createIndex({ "indexedInformations.osirisActionId": 1 }, { unique: true });
        await this.collection.createIndex({ "legalInformations.siret": 1 });
    }
}

const osirisEvaluationPort = new OsirisEvaluationPort();

export default osirisEvaluationPort;
