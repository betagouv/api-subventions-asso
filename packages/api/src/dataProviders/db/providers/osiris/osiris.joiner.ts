import { AggregationCursor } from "mongodb";
import db from "../../../../shared/MongoConnection";
import osirisActionPort from "./osiris.action.port";
import osirisRequestPort from "./osiris.request.port";
import OsirisRequestEntity from "../../../../modules/providers/osiris/entities/OsirisRequestEntity";
import OsirisActionEntity from "../../../../modules/providers/osiris/entities/OsirisActionEntity";

export type OsirisRequestWithActions = OsirisRequestEntity & { actions: OsirisActionEntity[] };

export class OsirisJoiner {
    applicationCollection = db.collection(osirisRequestPort.collectionName);

    private get joinPipeline() {
        return [
            {
                $lookup: {
                    from: osirisActionPort.collectionName,
                    localField: osirisRequestPort.joinIndexes.osirisActionPort,
                    foreignField: osirisActionPort.joinIndexes.osirisRequestPort,
                    as: "actions",
                },
            },
        ];
    }

    findAllCursor(): AggregationCursor<OsirisRequestWithActions> {
        return this.applicationCollection.aggregate(this.joinPipeline);
    }

    findByExerciseCursor(exercise: number): AggregationCursor<OsirisRequestWithActions> {
        return this.applicationCollection.aggregate([
            { $match: { "providerInformations.exercise": exercise } },
            ...this.joinPipeline,
        ]);
    }
}

const osirisJoiner = new OsirisJoiner();
export default osirisJoiner;
