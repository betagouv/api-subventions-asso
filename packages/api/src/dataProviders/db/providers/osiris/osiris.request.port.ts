import { FindOneAndUpdateOptions, ObjectId } from "mongodb";
import OsirisRequestEntity from "../../../../modules/providers/osiris/entities/OsirisRequestEntity";
import MongoPort from "../../../../shared/MongoPort";
import Siret from "../../../../valueObjects/Siret";
import Rna from "../../../../valueObjects/Rna";
import Siren from "../../../../valueObjects/Siren";

export class OsirisRequestPort extends MongoPort<OsirisRequestEntity> {
    collectionName = "osiris-requests";

    async createIndexes() {
        await this.collection.createIndex({ "providerInformations.uniqueId": 1 }, { unique: true });
        await this.collection.createIndex({ "providerInformations.osirisId": 1 });
        await this.collection.createIndex({ "legalInformations.rna": 1 });
        await this.collection.createIndex({ "legalInformations.siret": 1 });
    }

    public async add(osirisRequest: OsirisRequestEntity) {
        return this.collection.insertOne(osirisRequest);
    }

    /*
     * @deprecated
     * */
    public async update(osirisRequest: OsirisRequestEntity) {
        const options = { returnDocument: "after", includeResultMetadata: true } as FindOneAndUpdateOptions;
        const { _id, ...requestWithoutId } = osirisRequest;
        const updateRes = await this.collection.findOneAndUpdate(
            { "providerInformations.uniqueId": osirisRequest.providerInformations.uniqueId },
            { $set: requestWithoutId },
            options,
        );
        //@ts-expect-error -- mongo typing expects no metadata
        return updateRes?.value as OsirisRequestEntity;
    }

    public upsertOne(osirisRequest: OsirisRequestEntity) {
        const options = { upsert: true } as FindOneAndUpdateOptions;
        const { _id, ...requestWithoutId } = osirisRequest;
        return this.collection.updateOne(
            { "providerInformations.uniqueId": osirisRequest.providerInformations.uniqueId },
            { $set: requestWithoutId },
            options,
        );
    }

    public async findByMongoId(id: string): Promise<OsirisRequestEntity | null> {
        return this.collection.findOne({ _id: new ObjectId(id) });
    }

    public findByUniqueId(uniqueId: string) {
        return this.collection.findOne({
            "providerInformations.uniqueId": uniqueId,
        }) as unknown as Promise<OsirisRequestEntity | null>;
    }

    public findBySiret(siret: Siret) {
        return this.collection
            .find({
                "legalInformations.siret": siret.value,
            })
            .toArray();
    }

    public findByRna(rna: Rna) {
        return this.collection
            .find({
                "legalInformations.rna": rna.value,
            })
            .toArray();
    }

    public async findBySiren(siren: Siren) {
        return this.collection
            .find({
                "legalInformations.siret": new RegExp(`^${siren.value}\\d{5}`),
            })
            .toArray();
    }

    public cursorFindRequests(query = {}) {
        return this.collection.find(query);
    }
}

const osirisRequestPort: OsirisRequestPort = new OsirisRequestPort();
export default osirisRequestPort;
