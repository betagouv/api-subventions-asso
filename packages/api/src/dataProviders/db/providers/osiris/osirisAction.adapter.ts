import { WithId } from "mongodb";
import OsirisActionEntity from "../../../../modules/providers/osiris/entities/OsirisActionEntity";

export default class OsirisActionAdapter {
    public static toEntity(dbo: WithId<OsirisActionEntity>): OsirisActionEntity {
        return new OsirisActionEntity(dbo.indexedInformations, dbo.data, dbo.updateDate);
    }
}
