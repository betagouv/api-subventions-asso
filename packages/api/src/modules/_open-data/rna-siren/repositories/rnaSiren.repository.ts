import { Siren, Rna } from "dto";
import { ObjectId } from "mongodb";
import RnaSiren from ".././entities/RnaSirenEntity";
import MongoRepository from "../../../../shared/MongoRepository";

export class RnaSirenRepository extends MongoRepository<RnaSiren> {
    collectionName = "rna-siren";

    findById(id: ObjectId) {
        return this.collection.findOne({ _id: id });
    }

    findBySiren(siren: Siren) {
        return this.collection.findOne({ siren });
    }

    findByRna(rna: Rna) {
        return this.collection.findOne({ rna });
    }

    async create(entity: RnaSiren) {
        return !!(await this.collection.insertOne(entity));
    }

    async insertMany(entities: RnaSiren[]) {
        return this.collection.insertMany(entities, { ordered: false });
    }

    async cleanDuplicate() {
        const duplicates: ObjectId[] = [];

        await this.collection
            .aggregate(
                [
                    { $match: {} }, // Find all entities
                    {
                        $group: {
                            // Group by rna
                            _id: { rna: "$rna" },
                            dups: { $addToSet: "$_id" },
                            count: { $sum: 1 },
                        },
                    },
                    {
                        $match: {
                            // Keep just groups to have length > 1
                            count: { $gt: 1 },
                        },
                    },
                ],
                { allowDiskUse: true }, // For faster processing if set is larger
            )
            .forEach(doc => {
                doc.dups.shift(); // First element skipped for deleting
                duplicates.push(...doc.dups);
            });

        // Remove all duplicates in one go
        await this.collection.deleteMany({ _id: { $in: duplicates } });
    }

    async createIndexes() {
        await this.collection.createIndex({ rna: 1 });
        await this.collection.createIndex({ siren: 1 });
    }
}

const rnaSirenRepository = new RnaSirenRepository();

export default rnaSirenRepository;
