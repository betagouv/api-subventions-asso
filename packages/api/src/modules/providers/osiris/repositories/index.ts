import osirisRequestRepository from "./osiris.request.repository";
import osirisActionRepository from "./osiris.action.repository";
import osirisEvaluationRepository from "./osiris.evaluation.repository";

export { osirisRequestRepository, osirisActionRepository, osirisEvaluationRepository };
