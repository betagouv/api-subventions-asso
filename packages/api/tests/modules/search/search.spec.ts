import request = require("supertest");
import getUserToken from "../../__helpers__/getUserToken";
import { osirisRequestRepository } from "../../../src/modules/providers/osiris/repositories";
import associationNameRepository from "../../../src/modules/association-name/repositories/associationName.repository";
import OsirisRequestEntityFixture from "../providers/osiris/__fixtures__/entity";
import AssociationNameFixture from "../association-name/__fixtures__/entity";
import apiEntrepriseService from "../../../src/modules/providers/apiEntreprise/apiEntreprise.service";
import dauphinService from "../../../src/modules/providers/dauphin/dauphin.service";

const g = global as unknown as { app: unknown };

describe("/search", () => {
    beforeAll(() => {
        jest.spyOn(apiEntrepriseService, "getHeadcount").mockImplementation(async () => null);
        jest.spyOn(dauphinService, "getDemandeSubventionBySiret").mockImplementation(async () => []);
        jest.spyOn(dauphinService, "getDemandeSubventionBySiren").mockImplementation(async () => []);
    });

    describe("/associations/{input}", () => {
        beforeEach(async () => {
            Promise.all(AssociationNameFixture.map(fixture => associationNameRepository.create(fixture)));
        });

        it("should return an Association from its RNA", async () => {
            const response = await request(g.app)
                .get(`/search/associations/${AssociationNameFixture[0].rna}`)
                .set("x-access-token", await getUserToken())
                .set("Accept", "application/json");

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                result: [{ lastUpdate: expect.any(String) }]
            });
        });
        it("should return an Association from its Siren", async () => {
            const response = await request(g.app)
                .get(`/search/associations/${AssociationNameFixture[0].siren}`)
                .set("x-access-token", await getUserToken())
                .set("Accept", "application/json");

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                result: [{ lastUpdate: expect.any(String) }]
            });
        });
        it("should return an AssociationNameEntity from its name", async () => {
            const response = await request(g.app)
                .get(`/search/associations/${AssociationNameFixture[0].name}`)
                .set("x-access-token", await getUserToken())
                .set("Accept", "application/json");

            expect(response.statusCode).toBe(200);
            expect(response.body).toMatchSnapshot({
                result: [{ lastUpdate: expect.any(String) }]
            });
        });
    });
});
