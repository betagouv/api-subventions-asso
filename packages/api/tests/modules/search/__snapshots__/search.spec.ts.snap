// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`/search /associations/{input} should return an Association from its RNA 1`] = `
Object {
  "result": Array [
    Object {
      "lastUpdate": Any<String>,
      "name": "Association Name",
      "provider": "Provider Name",
      "rna": "W000000000",
      "siren": "000000001",
    },
  ],
  "success": true,
}
`;

exports[`/search /associations/{input} should return an Association from its Siren 1`] = `
Object {
  "result": Array [
    Object {
      "lastUpdate": Any<String>,
      "name": "Association Name",
      "provider": "Provider Name",
      "rna": "W000000000",
      "siren": "000000001",
    },
  ],
  "success": true,
}
`;

exports[`/search /associations/{input} should return an AssociationNameEntity from its name 1`] = `
Object {
  "result": Array [
    Object {
      "lastUpdate": Any<String>,
      "name": "Association Name",
      "provider": "Provider Name",
      "rna": "W000000000",
      "siren": "000000001",
    },
  ],
  "success": true,
}
`;

exports[`/search /etablissement/{siret} should return an association 1`] = `
Object {
  "etablissement": Object {
    "adresse": Array [
      Object {
        "last_update": "2022-01-01T00:00:00.000Z",
        "provider": "Osiris",
        "type": "object",
        "value": Object {},
      },
    ],
    "association": Object {
      "denomination_rna": Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "string",
          "value": "OSIRIS_ENTITY_FIXTURE",
        },
      ],
      "etablisements_siret": Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "object",
          "value": Array [
            "00000000000001",
          ],
        },
      ],
      "rna": Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "string",
          "value": "W000000001",
        },
      ],
      "siren": Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "string",
          "value": "000000000",
        },
      ],
    },
    "contacts": Array [
      Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "object",
          "value": Object {},
        },
      ],
    ],
    "demandes_subventions": Array [
      Object {
        "actions_proposee": Array [],
        "annee_demande": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "number",
          "value": 2022,
        },
        "contact": Object {
          "email": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
        },
        "date_commision": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "object",
          "value": "2022-01-01T00:00:00.000Z",
        },
        "dispositif": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "undefined",
        },
        "ej": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "string",
          "value": "EJ00001",
        },
        "montants": Object {
          "accorde": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
          "demande": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
          "propose": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
          "total": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
        },
        "pluriannualite": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "undefined",
        },
        "service_instructeur": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "undefined",
        },
        "siret": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "string",
          "value": "00000000000001",
        },
        "sous_dispositif": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "undefined",
        },
        "status": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "undefined",
        },
        "territoires": Array [],
        "versement": Object {
          "acompte": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
          "compensation": Object {
            "n-1": Object {
              "last_update": "2022-01-01T00:00:00.000Z",
              "provider": "OSIRIS",
              "type": "undefined",
            },
            "reversement": Object {
              "last_update": "2022-01-01T00:00:00.000Z",
              "provider": "OSIRIS",
              "type": "undefined",
            },
          },
          "realise": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
          "solde": Object {
            "last_update": "2022-01-01T00:00:00.000Z",
            "provider": "OSIRIS",
            "type": "undefined",
          },
        },
        "versementKey": Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "OSIRIS",
          "type": "string",
          "value": "EJ00001",
        },
        "versements": Array [],
      },
    ],
    "information_banquaire": Array [],
    "nic": Array [
      Object {
        "last_update": "2022-01-01T00:00:00.000Z",
        "provider": "Osiris",
        "type": "string",
        "value": "00001",
      },
    ],
    "representants_legaux": Array [
      Array [
        Object {
          "last_update": "2022-01-01T00:00:00.000Z",
          "provider": "Osiris",
          "type": "object",
          "value": Object {},
        },
      ],
    ],
    "siege": Array [
      Object {
        "last_update": "2022-01-01T00:00:00.000Z",
        "provider": "Osiris",
        "type": "undefined",
      },
    ],
    "siret": Array [
      Object {
        "last_update": "2022-01-01T00:00:00.000Z",
        "provider": "Osiris",
        "type": "string",
        "value": "00000000000001",
      },
    ],
    "versements": Array [],
  },
  "success": true,
}
`;
